### ISSUE TITLE
Improve performance through polymorphism

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#381]
Right now, many methods accept `IO[E, Unit]`.

For example:

```scala
IO.bracket / io.bracket
IO.bracket0 / io.bracket0
io.fork0
io.ensuring
io.bracketOnError
io.managed
io.onError
io.onInterrupt
io.onTermination
IO.supervised / io.supervised
Promise.bracket
```

This means that if you pass an `IO[E, A]`, you must first call `void` to discard the return value.

On the one hand, this forces a user to be explicit: they must know the return value will be discarded. This makes it harder to get order of parameters wrong (e.g. `io.bracket`). On the other hand, calling `void` involves an extraneous `flatMap` operation that has impacts on performance critical code.

To improve performance, we can relax the constraint that `Unit` be returned from many (perhaps all) of these methods, and instead accept `IO[E, Any]`. 

This should not involve adding any new functionality, just changing some type signatures. After they are changed, some `.void` calls can be eliminated inside ZIO to improve performance.

**Comment by kaishh** [comment]
IMHO code inside zio can use unsafe .asInstanceOfs to avoid the cost, however to decrease typesafety in user code to avoid one (synchronous) op is a bit of an overstretch in my opinion.

**Comment by jdegoes** [comment]
It’s still type safe. It changes this:

def ensuring(f: IO[Nothing, Unit])

To this:

def ensuring(f: IO[Nothing, Any])

For example.

It means you can pass it IO that return values.

The bad side is that in some APIs, it becomes more easy to misuse (e.g. swapping order of release and use in bracket).

An alternative would be to define ZUnit as an opaque type. You’d have to ‘voidz’ On tasks but it would be a no-op.

**Comment by mijicd** [comment]
I'd like to pick this one assuming nobody already started to work on it.

**Comment by RaasAhsan** [comment]
I prefer being precise about the types I work with. As an alternative to `Any`, we can also introduce a type parameter where we don't care about the type. 

```scala
def bracket[E1 >: E, B, C](release: A => IO[Nothing, C])(use: A => IO[E1, B]): IO[E1, B]
```

After this is implemented, I'm interested in seeing a benchmark on how much performance this will squeeze out. There's definitely a lot of fish to be caught in terms of performance, but we should be hesitant to make changes that could potentially compromise the API in unforeseen ways. Making this change is very easy and would retain backwards compatibility, but once it is made, it will be difficult to undo without breaking existing code.

**Comment by regiskuckaertz** [comment]
How about 
```
def ensuring(f: IO[Nothing, _])
```

**Comment by jdegoes** [comment]
@RaasAhsan Parametric polymorphism is equivalent to subtype polymorphism in this case.

That is, if you are passing `io: IO[Nothing, String]` into a function, then this:

```scala
def acceptParametric[A](io: IO[Nothing, A]): IO[Nothing, Unit]
```

is equivalent to this:


```scala
def acceptSubtype(io: IO[Nothing, Any]): IO[Nothing, Unit]
```

In both cases, the `accept` functions have no ability to know anything about the type returned by the `io` (assuming they are playing by Scalazzi rules).

It's not type unsafe. It would be type unsafe to _return_ an `IO[E, Any]`, because there's nothing the user can do with such a value, unless they do runtime type checking. But it's not type unsafe to _accept_ an `IO[E, Any]`.

The only difference between the parametric version and the subtype version is that Scala will consistently infer the subtype version, but may struggle in some cases with the parametric version.

@regiskuckaertz 

This construct:

```scala
def ensuring(f: IO[Nothing, _])
```

basically means:

```scala
def ensuring(f: IO[Nothing, A] forSome { type A } )
```

i.e. it's an existential type. It's another way of obscuring the type from the receiver of the method. However, as I understand it, these types are going away in Scala 3 so we should not rely on them in ZIO.

> After this is implemented, I'm interested in seeing a benchmark on how much performance this will squeeze out.

For sure. The current RTS continues to shrink and is evolving toward a "micro kernel" design in which most of the runtime is in IO, aside from a small bootstrapped portion. This makes it important to not add overhead unless there's a good reason to do so. In this particular case, I'm not sure, but I don't think that the overhead of `.void` really pays for itself.

**Comment by RaasAhsan** [comment]
@jdegoes Thanks for the explanation, that was very informative! When I had my original thought, I recognized it was similar to `Any` in some way, but didn't understand the complete relationship. I know the Scala compiler is very wonky with type inference, but do you have any compelling examples for why we would want one over the other in terms of utilizing parametric vs subtype polymorphism here?

**Comment by kaishh** [comment]
Wildcards are not going anywhere, only `forSome` existentials. Even simple kind `*` `forSome`s have a subtly different behavior than wildcards in my experience. Also, in my experience `_` inference is a lot better than `Any`, I don't have any examples, but it may be because wildcard does not rely on variance.

**Comment by jdegoes** [comment]
I agree I've never had issues with wildcard inference. 

[This issue](https://github.com/lampepfl/dotty/issues/5379) discusses the possibility of `_` going away as syntax for wildcards, possibly being replaced by `?`. However it seems this won't happen till Scala 3.2, in which case maybe we should use `_` instead of `Any`.
